### PR TITLE
Added message format checker

### PR DESCRIPTION
**Pull request by MrPointer**
This work adds support for a **Format Checker**, which accepts a message represented by a C-String and checks whether it fits the expected format of an NMEA message.

This work also comes with a handful of unit tests which should provide full coverage.
It's also well-documented.

@RonAmihai One thing I'd like to get your attention on is the way the tests are written.
I've left a `todo` note where a factory may be needed, but I couldn't decide on my own.

### CODE REVIEW

**Comment by MrPointer** [comment]
I've added missing coverage, which also caused to modify some checks to be even "smarter".

**Comment by RonAmihai** [comment]
Feedback about the code itself (without tests):

1. EVALID code should be returned instead of explicit 0
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L59

2. Unnecessary 'else-if' (first 'if' includes 'return' statement & 'else' doesn't improves readability anyways) 
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L87

3. Consider to name out parameters with 'out' prefix to improve readability 
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L93 

4. Use contants (defines) for offsets & length instead of explicit values
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L37
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L155
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/format/FormatChecker.c#L107
  
5. Use conteants (defines) for shifting values. They can be understood eventually but it is desirable to specify their vocation explicitly for readability 
https://github.com/takpit/proto-nmea/blob/d057163d9dfe2b12bcd23e42c52458a456142940/src/utility/GlobalUtils.c#L39

File: src/format/FormatChecker.c [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    errorCode = validateMessageType(rawMessage);
+    if (errorCode)
+        return errorCode;
+
+    errorCode = validateChecksumData(rawMessage, checksumDelimiterPosition);
+    if (errorCode)
+        return errorCode;
+
+    return 0;
```
Comment (RonAmihai) [inline src/format/FormatChecker.c:57]: EVALID code should be returned instead of explicit 0

File: src/format/FormatChecker.c [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+static int8_t validateMessageLengthInRange(size_t messageSize)
+{
+    if (messageSize <= MESSAGE_MIN_LENGTH)
+    {
+        if (messageSize < MESSAGE_MIN_LENGTH)
+            return -ESHORT_MESSAGE;
+        else
+            return -EMISSING_MESSAGE_DATA;
+    }
+    else if (messageSize > MESSAGE_MAX_LENGTH)
```
Comment (RonAmihai) [inline src/format/FormatChecker.c:73]: Unnecessary 'else-if' (first 'if' includes 'return' statement & 'else' doesn't improves readability anyways)

File: src/format/FormatChecker.c [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        return -ELONG_MESSAGE;
+
+    return EVALID;
+}
+
+static int8_t validateChecksumFormat(const char *message, size_t messageSize, size_t *checksumIdealIndex)
```
Comment (RonAmihai) [inline src/format/FormatChecker.c]: Consider naming out-parameters with 'out' prefix to improve readability

File: src/format/FormatChecker.c [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+int8_t validateFixedSizeMessageFormat(const char *rawMessage, size_t messageSize)
+{
+    int8_t errorCode = validateInitialMessageFormat(rawMessage);
+    if (errorCode)
+        return errorCode;
+
+    // Validate message ending fits protocol
+    if (rawMessage[messageSize - 2] != PROTOCOL_STOP_CHAR_1 ||
```
Comment (RonAmihai) [inline src/format/FormatChecker.c]: Use constants (defines) for offsets & length instead of explicit values

File: src/format/FormatChecker.c [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    // Should start with a checksum delimiter
+    if (message[*checksumIdealIndex] != PROTOCOL_CHECKSUM_DELIMITER)
+    {
+        if (strchr(message + *checksumIdealIndex, PROTOCOL_CHECKSUM_DELIMITER))
+            return -EMISSING_CHECKSUM_DATA;
+        else
+            return -EMISSING_CHECKSUM;
+    }
+
+    if (!isHex(message[*checksumIdealIndex + 1]) || !isHex(message[*checksumIdealIndex + 2]))
```
Comment (RonAmihai) [inline src/format/FormatChecker.c]: Use constants (defines) for offsets & length instead of explicit values

File: src/format/FormatChecker.c [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    return EVALID;
+}
+
+static int8_t validateChecksumData(const char *message, size_t checksumDelimiterIndex)
+{
+    char receivedChecksum[3] = {0};
+
+    // ToDo: Optimize?
+    memcpy(receivedChecksum, message + checksumDelimiterIndex + 1, 2);
```
Comment (RonAmihai) [inline src/format/FormatChecker.c]: Use constants (defines) for offsets & length instead of explicit values

File: src/utility/GlobalUtils.c [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+#endif
+
+//endregion
+
+int stringToHex(unsigned const char *hexString)
+{
+    int ret = 0;
+    while (*hexString && ret >= 0)
+    {
+        ret = (ret << 4) | hextable[*hexString++];
```
Comment (RonAmihai) [inline src/utility/GlobalUtils.c]: Use constants (defines) for shifting values. They can be understood eventually but it is desirable to specify their vocation explicitly for readability

**Review by RonAmihai — COMMENTED** [review]
Review of the code itself (without tests)

**Comment by RonAmihai** [comment]
As for the test - only two things to mention: 
1) Consider using constants (defines) also for string-append indexers. 
2) Factories would be nice addition whenever you mentioned.  

Other than that - the tests are written good overall.

File: src/format/FormatChecker.c (hunk 3 above)
Comment (MrPointer) [reply src/format/FormatChecker.c]: I disagree. The **out** modifier is mostly known to .NET developers (C# specifically), yet it's not very popular in C.
With that being said, I do agree that this function isn't very consistent with its' name, and that should be addressed - Either by separating code somehow or by renaming it.

File: src/format/FormatChecker.c (hunk 4 above)
Comment (MrPointer) [reply src/format/FormatChecker.c]: In this case we're merely checking the 2 last elements of the array, I don't think there's a need to name it somehow - Care to tell me if I'm wrong?

File: src/format/FormatChecker.c (hunk 5 above)
Comment (MrPointer) [reply src/format/FormatChecker.c]: Took a better approach by iterating over the checksum's length - That way it'll work even in extreme cases

**Pull request merged**